File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    unordered_map<int,int>mp;
    void isValid(TreeNode* root) {
        if(root==NULL) return;
        isValid(root->left);
        mp[root->val]++;
        isValid(root->right);
    }
    vector<int> findMode(TreeNode* root) {
        isValid(root);
        int maxi=0;
        for(auto it:mp){
            if(it.second>maxi) maxi=it.second;
        }
        vector<int>ans;
        for(auto it:mp){
            if(it.second==maxi) ans.push_back(it.first);
        }

        return ans;
    }
};

File: 0501-find-mode-in-binary-search-tree/0501-find-mode-in-binary-search-tree.js
var findMode = function(root) {
    // Initialize a Map to store the frequency of each value
    var frequencyMap = new Map();

    // Helper function to perform an in-order traversal and count the frequency of each value
    function isValid(node) {
        if (node === null) return;

        isValid(node.left); // Traverse left subtree
        frequencyMap.set(node.val, (frequencyMap.get(node.val) || 0) + 1); // Update frequency count in the map
        isValid(node.right); // Traverse right subtree
    }

    isValid(root); // Call the helper function to populate the frequency map

    let maxi = 0;
    // Find the maximum frequency encountered
    frequencyMap.forEach(value => {
        maxi = Math.max(maxi, value);
    });

    let ans = [];
    // Collect elements with the maximum frequency in the result list
    frequencyMap.forEach((value, key) => {
        if (value === maxi) {
            ans.push(key);
        }
    });

    return ans;
};
